pkg/controllers/machine_controller.go

```diff
@@ -268,35 +265,29 @@ func (r *MachineReconciler) startCheckAndEnqueueVolumeResize(ctx context.Context
 }
 
 func (r *MachineReconciler) startCheckAndEnqueueMachineState(ctx context.Context, log logr.Logger) {
-	wait.UntilWithContext(ctx, func(ctx context.Context) {
-		machines, err := r.machines.List(ctx)
-		if err != nil {
-			log.Error(err, "failed to list machines")
-			return
-		}
+	eventChan, err := r.libvirt.LifecycleEvents(ctx)
+	if err != nil {
+		log.Error(err, "failed to subscribe lifecycle events")
+		return
+	}
 
-		for _, machine := range machines {
-			if machine.DeletedAt != nil || !slices.Contains(machine.Finalizers, MachineFinalizer) {
-				continue
-			}
-			id := machine.ID
+	log.Info("subscribing lifecycle events")
 
-			state, err := r.getMachineState(id)
-			if err != nil {
-				if libvirt.IsNotFound(err) {
-					log.V(1).Info("Failed to retrieve domain. Requeueing", "machineID", id)
-					r.queue.AddRateLimited(id)
-				} else {
-					log.Error(err, "failed to get machine state", "machineID", id)
-				}
-				continue
+	for {
+		select {
+		case ev, ok := <-eventChan:
+			if !ok {
+				log.Error(errors.New("missing lifecycle event"), "lifecycle event channel is closed")
+				return
 			}
 
-			if state != machine.Status.State {
-				r.queue.AddRateLimited(id)
-			}
+			log.Info("requeue machine id " + ev.Dom.Name)
+			r.queue.AddRateLimited(ev.Dom.Name)
```

Comment:
Is it a good idea to pass all the events here? I am not sure how many of those events are coming in. Can't we create a `filterLibvirtEvents` helper method to catch only the events which are interesting for the machine lifecycle?

Reply:
it is same logic as virtlet has: https://github.com/onmetal/virtlet/blob/main/source/source.go#L104

Reply:
i looked at lifecycle events more deeply and for me every "subevents" look usable: https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainEventType
I will extend log message about event ID, so we can monitor it and we can prepare improvements later.
I would like extend your idea with another events: https://github.com/ironcore-dev/libvirt-provider/issues/192

Reply:
partially fixed